SSR-2540 Table and ReadOnlyTable tips (#1215)
kintone-labs/kintone-ui-component: docs/document/docs/guides/table-readonly-table-customization.md

---
id: table-readonly-table-customization
title: Table and readOnlyTable customization
sidebar_label: Table and readOnlyTable customization
---

## Overview
This article explains how to utilize and customize the Table, ReadOnlyTable component and the usage difference between them.<br>
We assume the following scenario:
1. Table and ReadOnlyTable usage difference
2. Table and ReadOnlyTable: Create simple one
3. Table: Set multiple components in one cell
4. Table: Update cell value depending on the other cell value change

> 5. Table: Table in table

Reviewer: How about unifying the number and title in the detailed section?

Author: How about new order ?
<img width="1194" alt="Screen Shot 2023-02-01 at 10 18 13" src="https://user-images.githubusercontent.com/48341495/215938477-c5ecf861-411c-4893-8791-05942a6a1003.png">

Reviewer: My comment was not good. I will explain it again.
I think the explanation order is ok.
I mean that "Overview > scenario" title and "Examples" title are different, so it will be easier to understand for users if it is unified.
How do you think?

Author: <img width="1020" alt="Screen Shot 2023-02-01 at 13 46 27" src="https://user-images.githubusercontent.com/48341495/215971384-ca4710f4-2c0c-4315-a9b2-f2a6dd6190d5.png">

is it ok with the above update for the scenario and examples?

Reviewer: Thanks!
Please choose from these two options.
- Only write Example code scenarios on scenario list (if this, need deleting "usage difference")
- Write full agenda on scenario list (if this, need adding "Understanding change event")

Author: Thank for these options.
I will update following option 1 (deleting "usage difference")

Reviewer: Got it!

## Components to use
- [Dropdown](../components/desktop/dropdown.md)
- [Table](../components/desktop/table.md)
- [Text](../components/desktop/text.md)
- [ReadOnlyTable](../components/desktop/readonly-table.md)
- [RadioButton](../components/desktop/radio-button.md)

## Table and ReadOnlyTable usage difference
Basically, the Table and ReadOnlyTable component have the same structure.

The ReadOnlyTable component allows the user to display a read-only mode table. This means that it only displays plain text that the user cannot manipulate.</br>

In contrast, the Table allows the user to manipulate the components displayed in the table cell such as changing data, adding and deleting records.

The biggest difference between Table and ReadOnlyTable components is the columns property of the table have a render property (`columns.render`) to specify which elements to display in the cell that the user can interact with. In addition, columns also have `actionButton` property (`table.actionButton`) to show/hide the add/remove row button.

## Understanding `change` event
It will be easier for you to customize the table when you understand the `change` event flow in the Table.

> >The table cell will listen `change` event from the component inside the cell and assign that new value to the field.</br>

Author: The "field" is columns's property. ( `columns.field` )
Example: We have ` const columns = [ { title: 'Country', field: 'country', render: renderDropdownCountry } ] `
When the dropdown inside country columns changed, the `new value` will be assigned to data (data[i].country = `new value`)

> >The table cell will listen `change` event from the component inside the cell and assign that new value to the field.</br>

Reviewer: Understood!
How about changing like this if it's correct?

```suggestion
>The table cell will listen `change` event from the component inside the cell and assign that new value to the corresponding `field` value.<br>
```

> >The table cell will listen `change` event from the component inside the cell and assign that new value to the field.</br>

Author: It's correct. I will commit your suggestion

> ### Create an App

Author: I want to explain how the table behaves when a data is changed
- create app
- prepare simple data
- display on UI
- modify change event

> ### Create an App

Reviewer: Got it! It's ok.

Create an app that includes a blank space field with the id "space".

### JavaScript and CSS Customization
When you import the UMD file of Kintone UI Component to the app, you can upload the JavaScript files by following these steps:<br>
You can see how to upload a file in the [Quick Start](../getting-started/quick-start.md).

### Prepare basic table
Suppose we have the following settings to display username with Text component:
```javascript
const renderName = (cellData) => {
  const text = new Kuc.Text({ value: cellData });
  return text;
};

const columns = [{ title: "Username", field: "username", render: renderName }];
const data = [{ username: "user1" }, { username: "user2" }];

const table = new Kuc.Table({columns, data});
const space = kintone.app.record.getSpaceElement('space');
space.appendChild(table);
```

### The display on UI

![render](assets/table-edit-text.gif)
When you edit the input in the first cell, the Text component will trigger `change` event with `detail.value = "user1 edited"`.

> The table then catch that value and assigns it back to the `username` field.

Reviewer: ```suggestion
The table then catches that value and assigns it back to the `username` field.
```

Now the data of the table will be:
```javascript
[{ username: "user1 edited" }, { username: "user2" }];
```

### Modify the event detail

> If you want to customize the value the table receives then you need to modify it in the `change` event before the table receives it.

Reviewer: Will it be ok with the sentence below?
```suggestion
If you want to customize the value that the table receives, you need to modify it in the `change` event before the table receives it.
```

Author: Yes, it's ok

> We also use the sample code above and add the following content to `renderName` function:

Reviewer: ```suggestion
We also use the sample code above and add the following content to the `renderName` function:
```

```javascript

> // table.js

Reviewer: Do we need this line??
```suggestion
```

Author: I think deleting it is ok. I just commented for the filename

...
const renderName = (cellData) => {
  const text = new Text({ value: cellData });

  // Modify the value before it bubble to table cell
  text.addEventListener('change', (event) => {
    event.detail.value = "modified value" // add any value you want set to username;
  });
  return text;
};
...
```

> When you edit the input in the first cell, The value received by the table will always be `"modified value"`

Reviewer: ```suggestion
When you edit the input in the first cell, the value received by the table will always be `"modified value"`
```

## Examples
### Basic Usage

#### ReadOnlyTable
![readonly-table](assets/readonly-table.png)

Display a read-only mode table.
<details>
  <summary>Show code</summary>

  ```js
const columns = [
    {
        title: 'Name',
        field: 'name',
    },
    {
        title: 'Gender',
        field: 'gender',
    },
    {
        title: 'Address',
        field: 'address',
    },
];

const data = [
    {
        name: 'John Brown',
        gender: 'male',
        address: 'osaka-japan',
    },
    {
        name: 'Jim Green',
        gender: 'female',
        address: 'tokyo-japan',
    },
    {
        name: 'Joe Black',
        gender: 'male',
        address: 'hochiminh-vietnam',
    },
];

const readOnlyTable = new Kuc.ReadOnlyTable({columns, data});
const space = kintone.app.record.getSpaceElement('space');
space.appendChild(readOnlyTable);
  ```
</details>

#### Table
![table](assets/table.png)

You can manipulate with the components displayed in the table.</br>

> Use the `columns.render` function to specify the component to be displayed in the cell. Please check `renderGender` and `renderAddress` functions in below code example:

Reviewer: ```suggestion
Use the `columns.render` function to specify the component to be displayed in the cell. Please check the `renderGender` and the `renderAddress` functions in the following code example:
```

<details>
  <summary>Show code</summary>

  ```js
// render gender column with dropdown
const renderGender = (cellData) => {
    const radioButton = new Kuc.RadioButton({
        items: [
        {
            label: 'Male',
            value: 'male',
        },
        {
            label: 'Female',
            value: 'female',
        },
        ],
        itemLayout: 'vertical',
        value: cellData,
    });

    return radioButton;
};

// render address column with dropdown
const renderAddress = (cellData) => {
    const country = cellData.split('-')[1];
    const dropdownCountry = new Kuc.Dropdown({
        items: [
        {

>             label: 'Việt Nam',

Reviewer: Please use English name. Please update image, too.
```suggestion
            label: 'Viet Nam',
```

Author: I updated the image

            value: 'vietnam',
        },
        {
            label: 'Japan',
            value: 'japan',
        },
        ],
        value: country,
    });

    return dropdownCountry;
};

const columns = [
    {
      title: 'Name',
      field: 'name',
    },
    {
      title: 'Gender',
      field: 'gender',
      render: renderGender,
    },
    {
      title: 'Address',
      field: 'address',
      render: renderAddress,
    },
];

const data = [
    {
        name: 'John Brown',
        gender: 'male',
        address: 'osaka-japan',
    },
    {
        name: 'Jim Green',
        gender: 'female',
        address: 'tokyo-japan',
    },
    {
        name: 'Joe Black',
        gender: 'male',
        address: 'hochiminh-vietnam',
    },
];

const table = new Kuc.Table({columns, data});
const space = kintone.app.record.getSpaceElement('space');
space.appendChild(table);
  ```
</details>

### Multiple components in one cell
![multi-components](assets/two-component-in-cell.png)

> You can put multiple component in one table cell using `columns.render` function.</br>

Reviewer: Please note that this example uses Table component.

```suggestion
You can put multiple components in one table cell using the `columns.render` function.<br>
```

> The example code below will display two dropdowns (city and country) in the same cell.

Reviewer: ```suggestion
The following example code will display two dropdowns (city and country) in the same cell.
```

<details>
  <summary>Show code</summary>

  ```js
  const renderAddress = (cellData, rowData) => {

>     // the format of cellData: "city-country";

Reviewer: Do we need ; mark?
```suggestion
    // the format of cellData: "city-country"
```

Author: it just a comment, we dont need ; mark.

          value: 'vietnam',
        },
        {
          label: 'Japan',
          value: 'japan',
        },
      ],
      value: country,
    });
    dropdownCountry.addEventListener('change', (event) => {
      const _city = rowData.address.split('-')[0];
      event.detail.value = `${_city}-${event.detail.value}`;
    });

    const container = document.createElement('div');
    container.style.display = 'flex';
    container.appendChild(dropdownCountry);
    container.appendChild(dropdownCity);

    return container;
  };

  const data = [
    {
      name: 'John Brown',
      gender: 'male',
      address: 'osaka-japan',
    },
    {
      name: 'Jim Green',
      gender: 'female',
      address: 'tokyo-japan',
    },
    {
      name: 'Joe Black',
      gender: 'male',
      address: 'hochiminh-vietnam',
    },
  ];

  const columns = [
    {
      title: 'Name',
      field: 'name',
    },
    {
      title: 'Address',
      field: 'address',
      render: renderAddress,
    },
  ];

const table = new Kuc.Table({columns, data});
const space = kintone.app.record.getSpaceElement('space');
space.appendChild(table);
  ```
</details>

### Dependent columns

> ![dependent-columns](assets/dependent-columns.gif)

Reviewer: Please change the image with the English version name of Viet Nam and Ho Chi Minh.
```suggestion
![dependent-columns](assets/dependent-columns.gif)

```

Author: I changed the image

> Update the city dropdown when the country dropdown changes.

Reviewer: ```suggestion
You can update the city dropdown when the country dropdown changes.
```

<details>
  <summary>Show code</summary>

  ```js

> // Each country will have corresponding cities.

Reviewer: ```suggestion
// Each country will have corresponding cities
```

const relatedData = {
    japan: [
        {label: 'Tokyo', value: 'tokyo'},
        {label: 'Osaka', value: 'osaka'},
    ],
    vietnam: [
        {label: 'Ha Noi', value: 'hanoi'},
        {label: 'Ho Chi Minh', value: 'hochiminh'},
    ],
};

const renderCity = (cellData, rowData) => {
    const dropdownCity = new Kuc.Dropdown({
      items: [
        {
          label: 'Tokyo',
          value: 'tokyo',
        },
        {
          label: 'Ho Chi Minh',
          value: 'hochiminh',
        },
      ],
      value: cellData,
    });

    // Logic update city when country column changed

>     lastRenderedCountryComponent.addEventListener('change', (e) => {
>       dropdownCity.items = relatedData[e.detail.value];
>       rowData.city = '';
>     });
> 
>     return dropdownCity;
>   };

Reviewer: ```suggestion
    lastRenderedCountryComponent.addEventListener('change', (event) => {
      dropdownCity.items = relatedData[event.detail.value];
      rowData.city = '';
    });

    return dropdownCity;
  };
```

  let lastRenderedCountryComponent;
  const renderCountry = (cellData) => {
    const dropdownCountry = new Kuc.Dropdown({
      items: [
        {
          label: 'Viet Nam',
          value: 'vietnam',
        },
        {
          label: 'Japan',
          value: 'japan',
        },
      ],
      value: cellData,
    });
    lastRenderedCountryComponent = dropdownCountry;
    return dropdownCountry;
  };

  const columns = [
    {
      title: 'Country',
      field: 'country',
      render: renderCountry,
    },
    {
      title: 'City',
      field: 'city',
      render: renderCity,
    },
  ];

  const data = [
    {
      country: 'japan',
      city: 'tokyo',
    },
    {
      country: 'vietnam',
      city: 'hochiminh',
    },
  ];

const table = new Kuc.Table({columns, data});
const space = kintone.app.record.getSpaceElement('space');
space.appendChild(table);
  ```
</details>

### Nested tables

> ![table-in-table](assets/table-in-table.png)

Reviewer: ```suggestion
![table-in-table](assets/table-in-table.png)

```

> Showing more detailed info of every row.

Reviewer: ```suggestion
You can show more detailed info on every row with nested tables.
```
